blocks/certifier.go

```diff
@@ -203,27 +214,50 @@ func (c *Certifier) RegisterForCert(ctx context.Context, lid types.LayerID, bid
 
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	c.createIfNeeded(lid, bid)
-	c.certifyMsgs[lid][bid].registered = true
-	return c.tryGenCert(ctx, logger, lid, bid)
+	info := c.createIfNeeded(lid, bid)
+	info.registered = true
+	return c.tryGenCert(ctx, logger, lid, bid, info)
 }
 
 // CertifyIfEligible signs the hare output, along with its role proof as a certifier, and gossip the CertifyMessage
 // if the node is eligible to be a certifier.
-func (c *Certifier) CertifyIfEligible(ctx context.Context, logger log.Log, lid types.LayerID, bid types.BlockID) error {
+func (c *Certifier) CertifyIfEligible(ctx context.Context, lid types.LayerID, bid types.BlockID) error {
 	if _, err := c.beacon.GetBeacon(lid.GetEpoch()); err != nil {
 		return errBeaconNotAvailable
 	}
-	// check if the node is eligible to certify the hare output
-	proof, err := c.oracle.Proof(ctx, lid, eligibility.CertifyRound)
+
+	var (
+		errsMu sync.Mutex
+		errs   error
+		eg     errgroup.Group
+	)
+	c.mu.Lock()
+	for _, s := range c.signers {
+		s := s
+		eg.Go(func() error {
+			if err := c.certifySingleSigner(ctx, s, lid, bid); err != nil {
+				errsMu.Lock()
+				errs = errors.Join(errs, fmt.Errorf("certifying block %v/%v by %s: %w", lid, bid, s.NodeID().ShortString(), err))
+				errsMu.Unlock()
+			}
+			return nil
+		})
+	}
+	c.mu.Unlock()
+	eg.Wait()
+	return errs
+}
+
+func (c *Certifier) certifySingleSigner(ctx context.Context, s *signing.EdSigner, lid types.LayerID, bid types.BlockID) error {
+	// check if the signer is eligible to certify the hare output
+	proof, err := c.oracle.Proof(ctx, s.VRFSigner(), lid, eligibility.CertifyRound)
 	if err != nil {
-		logger.With().Error("failed to get eligibility proof to certify", log.Err(err))
-		return err
+		return fmt.Errorf("getting eligibility proof: %w", err)
 	}
```

Comment:
you can use GenVRF here, and avoid useless error because you already got beacon before calling this method